### ISSUE TITLE
DoubleBagFTPS timeout in ruby

### ISSUE THREAD

**Issue by Gtar69** [bnix/double-bag-ftps#10]
I'm trying to use "double-bag-ftps" gem to connect local machine to ftp server but I met a issue => the code shown below

    p 'before ftp'
    ftps = DoubleBagFTPS.new
    ftps.passive = true
    ftps.ftps_mode = DoubleBagFTPS::EXPLICIT
    ftps.ssl_context = DoubleBagFTPS.create_ssl_context(:verify_mode =>    OpenSSL::SSL::VERIFY_NONE)
    ftps.connect('xx.xx.xx.xx', '990')
    ftps.login('user', 'password')

    p 'after ftp'
    p 'before filename'
    puts ftps.nlst.first
    p 'after filename'

the console output:

    "before ftp"
    DoubleBagFTPS:0x000000032ca378 @ftps_mode=:explicit, @ssl_context=#<OpenSSL::SSL::SSLContext:0x000000032ca030 @cert=nil, @key=nil, @client_ca=nil, @ca_file=nil, @ca_path=nil, @timeout=nil, @verify_mode=0, @verify_depth=nil, @renegotiation_cb=nil, @verify_callback=nil, @options=-2147482625, @cert_store=nil, @extra_chain_cert=nil, @client_cert_cb=nil, @tmp_dh_callback=nil, @session_id_context=nil, @session_get_cb=nil, @session_new_cb=nil, @session_remove_cb=nil, @servername_cb=nil, @npn_protocols=nil, @npn_select_cb=nil>, @mon_owner=nil, @mon_count=0, @mon_mutex=#<Mutex:0x000000032ca0a8>, @binary=true, @passive=true, @debug_mode=false, @resume=false, @sock=#<OpenSSL::SSL::SSLSocket:0x000000032c97c0 @io=#<TCPSocket:fd 7>, @context=#<OpenSSL::SSL::SSLContext:0x000000032ca030 @cert=nil, @key=nil, @client_ca=nil, @ca_file=nil, @ca_path=nil, @timeout=nil, @verify_mode=0, @verify_depth=nil, @renegotiation_cb=nil, @verify_callback=nil, @options=-2147482625, @cert_store=nil, @extra_chain_cert=nil, @client_cert_cb=nil, @tmp_dh_callback=nil, @session_id_context=nil, @session_get_cb=nil, @session_new_cb=nil, @session_remove_cb=nil, @servername_cb=nil, @npn_protocols=nil, @npn_select_cb=nil>, @sync_close=true, @hostname=nil, @eof=false, @rbuffer="", @sync=true, @callback_state=nil, @wbuffer="">, @logged_in=true, @open_timeout=nil, @read_timeout=60, @hostname="xx.xx.xx.xx", @last_response="200 PROT now Private.\n", @last_response_code="200", @ssl_session=#<OpenSSL::SSL::Session:0x000000032c9720>, @welcome="230 Login successful.\n">
    "after ftp"
    "before filename"
    /brabra/lib/double_bag_ftps.rb:83:in `initialize': Connection timed out - connect(2) (Errno::ETIMEDOUT)

login to ftps server works fine but it will stuck at "fetching first file name" in ftp server and result is timeout... No idea about this and very thanks.

**Comment by alwolfe** [comment]
We are having the exact same issue sees to have worked at some point.

**Comment by linkyndy** [comment]
Did you manage to find a solution to this? I'm having the exact problem...

**Comment by ranlei** [comment]
did you found a solution to this? i met the same problem.

**Comment by steveh** [comment]
I solved it by patching Net::Ftp to send EPSV instead of PASV. May be a quirk of your FTP server, I'm using vsftpd.

http://stackoverflow.com/questions/15653702/ruby-ftp-passive-mode-error
http://mble.github.io/blog/2014/12/24/ruby-net-ftp-and-extended-passive-mode/

**Comment by linkyndy** [comment]
Indeed, I've used a similar patch, but was looking for a more "official" approach to this.

**Comment by ealdent** [comment]
We have the same issue, and I agree that a more official patch would be ideal.

**Comment by StevenXL** [comment]
Interesting. I am using `vsftpd` as well, and having the same issue.
